### PR TITLE
[FLINK-3226] Translate logical joins to physical

### PR DESCRIPTION
**Pull request by vasia**
This PR contains the logical to physical translation for joins.
- Joins with equality conditions only do not generate a join function.
- Joins with equality and non-equality conditions evaluate the non-equality conditions inside a `FlatJoinFunction`.
- Joins without any equality conditions e.g. `in1.join.in2.where(a > b)` are currently not supported. We could consider supporting those later by generating a cross function.

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -148,16 +148,24 @@ class CodeGenerator(
...
         val inputTypeTerm = boxedTypeTermForTypeInfo(input1)
         ("Object map(Object _in1)",
-          s"$inputTypeTerm $input1Term = ($inputTypeTerm) _in1;")
+          List(s"$inputTypeTerm $input1Term = ($inputTypeTerm) _in1;"))
       }
 
+      // FlatJoinFunction
+      else if (clazz == classOf[FlatJoinFunction[_,_,_]]) {
+        val inputTypeTerm1 = boxedTypeTermForTypeInfo(input1)
+        val inputTypeTerm2 = boxedTypeTermForTypeInfo(input2.get) //TODO check this
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenerator.scala]: Why don't you call `getOrElse` here and throw an CodeGenException if input2 is not present? Then we can get rid of the TODO.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMap.scala [hunk 2]
```diff
@@ -63,7 +63,7 @@ class DataSetMap(
       config: TableConfig,
       expectedType: Option[TypeInformation[Any]])
     : DataSet[Any] = {
-    val inputDataSet = input.asInstanceOf[DataSetRel].translateToPlan(config)
+    val inputDataSet = input.asInstanceOf[DataSetRel].translateToPlan(config, expectedType)
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetMap.scala]: You cannot forward the expected type, since you don't know what the previous operator does.
E.g. if the expected type is `Tuple2` but the previous operator outputs records with one field.
Why did you change this call?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataset/DataSetJoinRule.scala [hunk 3]
```diff
@@ -39,18 +49,80 @@ class DataSetJoinRule
...
+    }
+  }
+
+  def getJoinFunction(join: FlinkJoin, joinInfo: JoinInfo):
+      ((TableConfig, TypeInformation[Any], TypeInformation[Any], TypeInformation[Any]) =>
+      FlatJoinFunction[Any, Any, Any]) = {
+
+    if (joinInfo.isEqui) {
+      // only equality condition => no join function necessary
+      null
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataset/DataSetJoinRule.scala]: In general, `null` is not very welcome in Scala. Could you return a `FlatJoinFunction` (containing only a `ConverterResultExpression`) here too? We can then get rid of the `Tuple2RowMapper` and support any type as output type of the join operation.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataset/DataSetJoinRule.scala [hunk 4]
```diff
@@ -39,18 +49,80 @@ class DataSetJoinRule
...
+    else {
+      val func = (
+        config: TableConfig,
+        leftInputType: TypeInformation[Any],
+        rightInputType: TypeInformation[Any],
+        returnType: TypeInformation[Any]) => {
+
+      val generator = new CodeGenerator(config, leftInputType, Some(rightInputType))
+      val condition = generator.generateExpression(join.getCondition)
+      val body = {
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataset/DataSetJoinRule.scala]: Unnecessary brackets?

**Comment by vasia** [comment]
Thanks for the review @twalthr! I've addressed your comments.

**Comment by twalthr** [comment]
Thanks for the changes. Looks good to merge ;-)

**Comment by vasia** [comment]
Great, merging this also.
